Added Android Hardware Buffer Extension Support

The-O-King opened the pull request:
## Description

Added support for the VK_ANDROID_EXTERNAL_MEMORY_ANDROID_HARDWARE_BUFFER
device extension on Android

Created necessary hooks for vkGetAndroidHardwareBufferPropertiesANDROID
and vkGetMemoryAndroidHardwareBufferANDROID

The-O-King committed:
Added Android Hardware Buffer Extension Support

Added support for the VK_ANDROID_EXTERNAL_MEMORY_ANDROID_HARDWARE_BUFFER
device extension on Android
*Created necessary hooks for vkGetAndroidHardwareBufferPropertiesANDROID
and vkGetMemoryAndroidHardwareBufferANDROID

baldurk reviewed (changes requested):
It seems to me like this implementation of the extension is incomplete, it's missing any handling of the new structs that can be chained onto `pNext`s when this extension is enabled.

The-O-King committed:
Added Struct Handling/Serialization

Added initial support for these structs:
*VkAndroidHardwareBufferPropertiesANDROID
*VkMemoryGetAndroidHardwareBufferInfoANDROID
*VkAndroidHardwareBufferFormatPropertiesANDROID
*VkExternalFormatANDROID
*VkAndroidHardwareBufferUsageANDROID
*VkImportAndroidHardwareBufferInfoANDROID

The-O-King commented:
I added support for the structs defined in the extension's documentation, although since this is my first time adding support for something like this in renderdoc I'm not sure if I missed anything

One thing I didn't complete yet was adding support for serializing the AHardwareBuffer struct that is passed into something like VkImportAndroidHardwareBufferInfoANDROID, do you have any documentation/examples in the code I could look at for doing something like this?

Thanks!

baldurk commented on renderdoc/driver/vulkan/vk_serialise.cpp:
```diff
@@ -9319,3 +9339,114 @@ INSTANTIATE_SERIALISE_TYPE(VkSurfaceCapabilitiesFullScreenExclusiveEXT);
+  SerialiseNext(ser, el.sType, el.pNext);
+
+  SERIALISE_MEMBER(allocationSize);
+  SERIALISE_MEMBER_TYPED(uint32_t, memoryTypeBits);
```
You don't need `SERIALISE_MEMBER_TYPED` here, `memoryTypeBits` is already a `uint32_t`.

baldurk commented on renderdoc/driver/vulkan/vk_serialise.cpp:
```diff
@@ -9319,3 +9339,114 @@ INSTANTIATE_SERIALISE_TYPE(VkSurfaceCapabilitiesFullScreenExclusiveEXT);
+            el.sType == VK_STRUCTURE_TYPE_IMPORT_ANDROID_HARDWARE_BUFFER_INFO_ANDROID);
+  SerialiseNext(ser, el.sType, el.pNext);
+
+  // SERIALISE_MEMBER(buffer);
```
I'd treat this the same way as `VkImportMemoryWin32HandleInfoKHR` does, cast the opaque buffer pointer to uint64_t to serialise the raw pointer value purely for informational purposes. We can't do anything with it on replay anyway, so just serialising it as an integer gives us something. Set it to `NULL` on replay because we'll strip the struct.

baldurk reviewed (changes requested):
There's handling needed in `vk_next_chains.cpp`, which is particularly important if you plan on handling this by stripping the structs on replay the same way as other external memory implementations.

Speaking of which you'll also need to extend the handling in `vkCreateImage` where it returns the worst case memory requirements of external and non-external images that have an import `pNext` struct. Do you know if there are memory types which are external only? That has been a problem in the past with the general external memory extensions.

The-O-King committed:
Handle pNext Chain for Android External Memory

Updated vk_next_chains.cpp to handle serializing of the new structs

Updated VkCreateInfo function to check for Android External Buffer

Fixes to vk_serialize.cpp to serialize structs properly

The-O-King commented:
I updated the PR based on your comments!

For your question about external memory - are you referring to memory types that can only be used to allocate an AHardwareBuffer? I'm not entirely sure I will have to ask/look around and see. Also let me know if I misunderstood the question!

baldurk commented:
Yes that's what I'm wondering about. If there's a memory type - say 5 - which is *only* available when importing a `AHardwareBuffer` and *can't* be used without one, then we can't replay that memory bind because we'd be binding to an invalid memory type. The code in `vkCreateImage` tries to avoid this by intersecting the queried memory types between the external and non-external image, but if there's no intersection at all (i.e. no memory types which can be used both with and without the `AHardwareBuffer`) then it's impossible to capture and replay correctly.

The-O-King commented:
So after asking around a bit it seems the answer to the question is entirely implementation dependent; so it is possible that there will be a memory type that can only be used for AHardwareBuffer

The scenario where the external image cannot be replayed seems to already be taken care of in the code, is there any additional things needed to be done to handle it from what is already there?

baldurk commented:
The problem is that such a scenario _isn't_ already taken care of. There's a warning printed when it's detected but then the code continues on and any such capture will then hit undefined behaviour on replay and most likely crash. I kind of hoped this would never happen.

I'm worried about giving people the false impression that this extension can be supported when actually on their implementation it can't. It's not very helpful to whitelist an extension if any use of it leads to a guaranteed failure. Unfortunately I can't see any way to query for this or detect it in advance and only conditionally report support for the extension (i.e. don't report the extension as supported if we can't capture and replay it).

Is this something an ISV is explicitly wanting, or was this a speculative implementation? I ask because the initial commit was non-functional so if this isn't something anyone wants yet maybe we can defer the problem until there's at least some known use case.

The-O-King commented:
Ah I see, this was implemented as we were running into issues internally where VkDevice creation was failing due to requesting the VK_ANDROID_EXTERNAL_MEMORY_ANDROID_HARDWARE_BUFFER extension, which causes applications being run through RenderDoc to crash on startup 

I had a dummy application setup that used the features of the extension as a test, so the main goal was to avoid the crash on startup, but I figured it might be good to try to implement it similar to how the Win32 external memory extensions were setup

baldurk commented:
I'm not sure I understand, you mean the application was enabling the extension without properly first checking to see if it was available? Or the application did require the extension and so failed when it wasn't available?

The-O-King commented:
The application did require the extension since the point of the application was to show applying an image filter to an Android created bitmap image, and VkDeviceCreate failed with VK_ERROR_EXTENSION_NOT_PRESENT

baldurk commented:
Ah ok, I misunderstood what you were getting at. Generally applications enumerate the available extensions so they can see exactly which extensions are and aren't supported and either gracefully fall back or at least identify the missing extension, but the vkCreateDevice error return is one way to do things.

I'll need to think about this. As I say whitelisting or implementing the extension won't make things any better for that application if it then just crashes on replay. At least having the extension not available is clear about what's wrong, having a mysterious replay crash would be worse.

baldurk commented:
Having thought about it, on balance between two unappealing choices I think merging this is slightly better. Unlike some other extensions this is one where applications are almost certainly going to have no fallback possible since it's not a rendering feature that can easily be turned off. For applications/hardware where the image can be replayed as non-external then this is obviously a win. For cases where this extension will be impossible to replay they do get a rather confusing experience, but in practice they were doomed either way. The only real loser is applications that do have a fallback if this extension isn't available and are running on hardware where it can't be replayed, but it's possible that there is no-one is in that precise situation.

Ideally the extension would have been structured such that this capability could be queried for in advance (or better yet, guaranteed by the spec) and so the extension could be properly hidden only when it can't be captured and replayed, but that ship has sailed.

What I'm going to do is add something to the memory-requirement checker on replay, such that if memory requirements are not satisfied on replay and the image is external then we'll explicitly call it out. The warning at capture time is unfortunately going to be easy to miss by the time the failure actually happens on replay.
